src/richie/apps/courses/admin.py

```diff
@@ -114,6 +165,60 @@ def has_module_permission(self, request):
         """
         return False
 
+    def get_form(self, request, obj=None, **kwargs):
+        """Add request to form so we can validate fields according to the user."""
+        form = super().get_form(request, obj=obj, **kwargs)
+        form.request = request
+        return form
+
+    def get_queryset(self, request):
+        """Exclude public course runs from admin views."""
+        user = request.user
+
+        qs = super().get_queryset(request)
+
+        if user.is_superuser or not getattr(settings, "CMS_PERMISSION", False):
+            return qs.filter(draft_course_run__isnull=True)
+
+        return qs.filter(
+            django_models.Q(direct_course__extended_object__pagepermission__user=user)
+            | django_models.Q(
+                direct_course__extended_object__pagepermission__group__user=user
+            ),
+            direct_course__extended_object__pagepermission__can_change=True,
+            draft_course_run__isnull=True,
+        )
+
+    def has_view_permission(self, request, obj=None):
+        """Allow view only if the user is allowed to change the related course page."""
+        if obj and getattr(settings, "CMS_PERMISSION", False):
```

Comment:
What do you think to move this check into a function ?
```python
if not user_can_change_course_page: return False
return super().has_view_permission(request, obj=obj)
```

Reply:
I thought the same but there are 2 levels of ifs so adding a function didn't save much... Actually `age_permissions.user_can_change_page` is already a shared function call.

Reply:
Okay. In fact we save space but lose in readability.